FinniversKit/Sources/Components/FrontPageShelfView/FrontPageShelfView.swift

```diff
@@ -0,0 +1,189 @@
+import UIKit
+
+public protocol FrontPageShelfViewDataSource {
+    func frontPageShelfView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath, withItem item: AnyHashable) -> UICollectionViewCell?
+    func frontPageShelfView(cellClassesIn collectionView: UICollectionView) -> [UICollectionViewCell.Type]
+    func datasource(forSection section: FrontPageShelfView.Section) -> [AnyHashable]
+}
+
+public protocol FrontPageShelfDelegate: AnyObject {
+    func frontPageShelfView(_ collectionView: UICollectionView, didSelectItem item: AnyHashable)
+}
+
+public class FrontPageShelfView: UIView {
+    typealias Datasource = UICollectionViewDiffableDataSource<Section, AnyHashable>
+    typealias Snapshot = NSDiffableDataSourceSnapshot<Section, AnyHashable>
+    
+    public enum Section: CaseIterable {
+        case savedSearch
+        case recentlyFavorited
+    }
+    
+    private var collectionViewDatasource: Datasource!
+    private var items: [Section: [AnyHashable]] = [:]
+    private var shelfDatasource: FrontPageShelfViewDataSource
+    public weak var shelfDelegate: FrontPageShelfDelegate?
+    
+    private var compositionalLayout: UICollectionViewCompositionalLayout {
+        let layout = UICollectionViewCompositionalLayout { sectionIndex, environment in
+            let section = Section.allCases[sectionIndex]
+            switch section {
+            case .savedSearch:
+                if self.items[section, default: []].isEmpty {
+                    fallthrough
+                }
+                return self.savedSearchLayout
+            case .recentlyFavorited: return self.favoriteLayout
+            }
+        }
+        
+        let config = UICollectionViewCompositionalLayoutConfiguration()
+        config.interSectionSpacing = 20
+        layout.configuration = config
+        return layout
+    }
+    
+    private lazy var collectionView: UICollectionView = {
+        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: compositionalLayout)
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        collectionView.delegate = self
+        return collectionView
+    }()
+    
+    public init(withDatasource datasource: FrontPageShelfViewDataSource) {
+        self.shelfDatasource = datasource
+        super.init(frame: .zero)
+        setup()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+// MAKR: - Public method
+public extension FrontPageShelfView {
+    func removeItem(_ item: AnyHashable) {
+        var snapshot = collectionViewDatasource.snapshot()
+        snapshot.deleteItems([item])
+        collectionViewDatasource.apply(snapshot, animatingDifferences: true)
+    }
+    
+    func reloadShelf() {
+        collectionViewDatasource = makeDatasource()
+        applySnapshot()
```

Comment:
Just curious, why make the dataSource again? Can't we just apply a new snapshot with the same dataSource? 🤔